modules/i2c/slavex.inc.c

```diff
@@ -6,7 +6,217 @@
  * The X_(p,s) macro must be defined before including.
  * It is automatically undefined at the end of this file.
  */
+#include <avr/interrupt.h>
 
-#error Slave mode not supported yet
+#define I2CX(s) X_(I2C,s)
+#define i2cX(s) X_(i2c,s)
+#define i2cs X_(i2cs,)
+#define TWIX X_(TWI,)
+#define twiX(s) X_(TWI,s)
 
+i2cs_t *i2csX;
+
+void i2cX(_init)(void)
+{
+}
+
+
+/*! \brief Initialize the I2C module. 
+ */
+void i2cX(s_init)(i2cs_t *p,
+                  uint8_t address,
+                  void (*process_data_function)(void)) {
+  i2csX = p;
+  i2csX->process_data = process_data_function;
+  i2csX->bytes_received = 0;
+  i2csX->bytes_transmit = 0;
+  i2csX->status = I2CS_STATUS_READY;
+  i2csX->result = I2CS_RESULT_UNKNOWN;
+
+  TWIX.SLAVE.CTRLA = (I2C_INTLVL<<TWI_SLAVE_INTLVL_gp) |
+                     TWI_SLAVE_DIEN_bm |
+                     TWI_SLAVE_APIEN_bm |
+                     TWI_SLAVE_ENABLE_bm;
+  TWIX.SLAVE.ADDR = (address<<1);
+}
+
+/*! \brief I2C transaction finished function.
+ *
+ *  Prepares module for new transaction.
+ *
+ *  \param i2c    The i2cs_t struct instance.
+ *  \param result The result of the transaction.
+ */
+void i2cX(_transaction_finished(uint8_t result)) {
```

Comment:
`i2cX` closing bracket should be placed after the function name: `void i2cX(_transaction_finished)(uint8_t result)`.
Also, internal methods should be declared `static`.

Other occurrences below.